File: app/src/main/java/ru/mrlargha/stemobile/tools/SubstitutionsSort.java
package ru.mrlargha.stemobile.tools;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

import ru.mrlargha.stemobile.data.model.Substitution;

public class SubstitutionsSort {
    public static ArrayList<Substitution> moveOldDatesToEnd(ArrayList<Substitution> substitutions,
                                                            Date referenceDate) {
        ArrayList<Substitution> actualSubstitutions = new ArrayList<>();
        ArrayList<Substitution> irrelevantSubstitutions = new ArrayList<>();
        Calendar reference = Calendar.getInstance();
        reference.setTime(referenceDate);
        reference.set(Calendar.HOUR_OF_DAY, reference.getActualMinimum(Calendar.HOUR_OF_DAY));
        reference.set(Calendar.MINUTE, reference.getActualMinimum(Calendar.MINUTE));
        reference.set(Calendar.SECOND, reference.getActualMinimum(Calendar.SECOND));
        reference.set(Calendar.MILLISECOND, reference.getActualMinimum(Calendar.MILLISECOND));

        Date clearedReferenceDate = reference.getTime();

        for (Substitution substitution :
                substitutions) {
            if (clearedReferenceDate.before(substitution.getSubstitutionDate())
                    || clearedReferenceDate.equals(substitution.getSubstitutionDate())) {
                actualSubstitutions.add(substitution);
            } else {
                irrelevantSubstitutions.add(substitution);
            }
        }

        actualSubstitutions.addAll(irrelevantSubstitutions);

        return actualSubstitutions;
    }
}

File: app/src/main/java/ru/mrlargha/stemobile/tools/SubstitutionsSort.kt
package ru.mrlargha.stemobile.tools

import ru.mrlargha.stemobile.data.model.Substitution
import java.util.*

object SubstitutionsSort {
    fun moveOldDatesToEnd(substitutions: ArrayList<Substitution>,
                          referenceDate: Date?): ArrayList<Substitution> {
        val actualSubstitutions = ArrayList<Substitution>()
        val irrelevantSubstitutions = ArrayList<Substitution>()
        val reference = Calendar.getInstance()
        reference.time = referenceDate
        reference[Calendar.HOUR_OF_DAY] = reference.getActualMinimum(Calendar.HOUR_OF_DAY)
        reference[Calendar.MINUTE] = reference.getActualMinimum(Calendar.MINUTE)
        reference[Calendar.SECOND] = reference.getActualMinimum(Calendar.SECOND)
        reference[Calendar.MILLISECOND] = reference.getActualMinimum(Calendar.MILLISECOND)
        val clearedReferenceDate = reference.time
        for (substitution in substitutions) {
            if (clearedReferenceDate.before(substitution.substitutionDate)
                    || clearedReferenceDate == substitution.substitutionDate) {
                actualSubstitutions.add(substitution)
            } else {
                irrelevantSubstitutions.add(substitution)
            }
        }
        actualSubstitutions.addAll(irrelevantSubstitutions)
        return actualSubstitutions
    }
}